# WebPSSM
[WebPSSM](https://webpssm.fredhutch.org) is a bioinformatic tool for predicting HIV-1 coreceptor usage from amino acid or nucleotide sequences of the third variable loop (V3) of the envelope gene. 
When a nucleotide sequence is entered, it will be translated to amino acid sequence first. If a nucleotide sequence contains ambiguous bases, it will be translated to all possible amino acid sequences. 
The original description of the method can be found [here](https://journals.asm.org/doi/10.1128/jvi.77.24.13376-13388.2003).

## Citation

  - Jensen, M. A., F.-S. Li, A. B. van 't Wout, D. C. Nickle, D. Shriner, H.-X. He, S. McLaughlin, R. Shankarappa, J. B. Margolick, and J. I. Mullins. 2003. Improved coreceptor usage prediction and genotypic monitoring of R5-to-X4 transition by motif analysis of HIV-1 env V3 loop sequences. [Journal of Virology 77: 13376-13388](http://jvi.asm.org/cgi/reprint/77/24/13376).
  - Brumme, Z. L., W. W. Dong, B. Yip, B. Wynhoven, N. G. Hoffman, R. Swanstrom, M. A. Jensen, J. I. Mullins, R. S. Hogg, J. S. Montaner, and P. R. Harrigan. 2004. Clinical and immunological impact of HIV envelope V3 sequence variation after starting initial triple antiretroviral therapy. [AIDS 18: F1-F9](https://journals.lww.com/aidsonline/fulltext/2004/03050/clinical_and_immunological_impact_of_hiv_envelope.1.aspx).
  - Jensen, M. A., M. Coetzer, A. B. van 't Wout, L. Morris, and J. I. Mullins. 2006. A reliable phenotype predictor for human immunodeficiency virus type 1 subtype C based on Envelope V3 sequences. [Journal of Virology 80: 4698-4704](http://jvi.asm.org/cgi/reprint/80/10/4698).

## Contact

For any questions, bugs and suggestions, please send email to cohnlabsupport@fredhutch.org and include a few sentences describing, briefly, the nature of your questions and include contact information.
